synapse/handlers/room_member.py

```diff
@@ -881,11 +902,11 @@ async def update_membership_locked(
             if action == "kick":
                 raise AuthError(403, "The target user is not in the room")
 
-        is_host_in_room = await self._is_host_in_room(state_before_join)
+        is_host_in_room = await self._is_host_in_room(partial_state_before_join)
```

Comment:
I think knocking (line +1052) is going to be broken in that we can end up sending a knock event from a user with a ban we don't know about.
We can't exactly await full state, since we're holding the room membership lock. Any thoughts on what we should do in this case?

Reply:
It's a bit different. Previously, we would do a remote /make_knock+/send_knock and the remote homeserver would tell us to go away we were banned.
Now, since we are in the room but do not have full state, we can add a local knock into the DAG that would otherwise be rejected and cite it as a prev/auth event.

Reply:
Hmm. Would it be worth requiring knocks to go over federation during the resync? (Not necessarily in this PR)

Reply:
I'm not wholly convinced the remote knock code is suitable. It persists the knock event as an outlier. I'm not sure what happens if we then receive the knock event over federation.

Reply:
I'm inclined to land this as-is and kick any knocking cans down the road.